### PR TITLE
BlockMax WAND

### PR DESCRIPTION
**Pull request by amourao**
### What's being changed:

- New inverted format that:
    - Writes docs ids and term frequencies in blocks
    - Compresses doc ids and frequencies
    - Writes tombstones to a separate area of the file
- Use BlockMax WAND to make search faster

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### Cross-functional impact

- [x] This change requires public documentation (weaviate-io) to be updated. Check the box to automatically create a corresponding issue.
    - Score values will change slightly when searching on multiple properties, due to using individual prop statistics instead of global statistics.
- Does it require a change in the client libraries? If yes, please check the boxes for the affected client libraries.
    - [ ] Python (weaviate-python-client)
    - [ ] JavaScript/TypeScript (typescript-client)
    - [ ] Go (weaviate-go-client)
    - [ ] Java (java-client)

### CODE REVIEW

File: adapters/repos/db/lsmkv/strategies_map.go [hunk 1]
```diff
@@ -211,6 +211,31 @@ func (kv MapPair) EncodeBytes(buf []byte) error {
...
 }
 
+func (kv MapPair) EncodeBytesInverted(buf []byte) error {
+	if len(buf) != invPayloadLen {
+		return errors.Errorf("buffer has size %d, but MapPair has size %d",
+			len(buf), invPayloadLen)
+	}
+
+	// make sure the 2 byte length indicators will never overflow:
+	if len(kv.Key) >= math.MaxUint16 {
```
Comment (jeroiraz) [inline adapters/repos/db/lsmkv/strategies_map.go:221]: this key validation seems rare

File: adapters/repos/db/lsmkv/strategies_map.go [hunk 2]
```diff
@@ -211,6 +211,31 @@ func (kv MapPair) EncodeBytes(buf []byte) error {
...
+		return errors.Errorf("mapCollection key must be smaller than %d",
+			math.MaxUint16)
+	}
+
+	if len(kv.Value) >= math.MaxUint16 {
```
Comment (jeroiraz) [inline adapters/repos/db/lsmkv/strategies_map.go:226]: this value len validation seems rare

File: adapters/repos/db/lsmkv/strategies_map.go [hunk 3]
```diff
@@ -249,6 +274,31 @@ func (kv MapPair) Bytes() ([]byte, error) {
...
+		return nil, errors.Errorf("mapCollection value must be smaller than %d",
+			math.MaxUint16)
+	}
+
+	out := bytes.NewBuffer(nil)
+
+	if _, err := out.Write(kv.Key); err != nil {
```
Comment (jeroiraz) [inline adapters/repos/db/lsmkv/strategies_map.go:291]: key length is not encoded but `FromBytesInverted` seems to be expecting a fixed size key of 8 bytes length. If that's the case then the key len validation in this method could validate exactly that

File: adapters/repos/db/lsmkv/strategies_map.go [hunk 4]
```diff
@@ -324,6 +374,31 @@ func (kv *MapPair) FromBytesReusable(in []byte, keyOnly bool) error {
...
+	// mmaped memory being removed from a completed compaction), however this is
+	// the wrong place. By the time we are in this method, we can no longer
+	// control the memory safety of the "in" argument. Thus, such a copy must
+	// happen at a much earlier scope when a lock is held that protects against
+	// removing the segment. Such an implementation can now be found in
+	// segment_collection_strategy.go as part of the *segment.getCollection
+	// method. As a result all memory used here can now be considered read-only
+	// and is safe to be used indefinitely.
+
+	kv.Key = in[read : read+8]
```
Comment (jeroiraz) [inline adapters/repos/db/lsmkv/strategies_map.go:391]: if may be worth validating len(in) >= 8

File: adapters/repos/db/shard_write_put.go [hunk 5]
```diff
@@ -523,9 +524,16 @@ func (s *Shard) updateInvertedIndexLSM(object *storobj.Object,
 	}
 
 	before := time.Now()
-	if err := s.extendInvertedIndicesLSM(propsToAdd, nilpropsToAdd, status.docID); err != nil {
-		return fmt.Errorf("put inverted indices props: %w", err)
+	if prevObject != nil && os.Getenv("USE_INVERTED_SEARCHABLE") == "true" {
```
Comment (jeroiraz) [inline adapters/repos/db/shard_write_put.go:527]: could you add a brief comment describing this change? I'm wondering why props is used when `prevObject != nil && os.Getenv("USE_INVERTED_SEARCHABLE") == "true"` holds

File: adapters/repos/db/shard_write_inverted_lsm.go [hunk 6]
```diff
@@ -87,8 +88,18 @@ func (s *Shard) addToPropertyValueIndex(docID uint64, property inverted.Property
 		if bucketValue == nil {
 			return errors.Errorf("no bucket searchable for prop '%s' found", property.Name)
 		}
+		propLen := float32(0)
 
-		propLen := float32(len(property.Items))
+		if os.Getenv("USE_INVERTED_SEARCHABLE") == "true" || os.Getenv("USE_NEW_PROPLENGTH") == "true" {
```
Comment (jeroiraz) [inline adapters/repos/db/shard_write_inverted_lsm.go:93]: I'd use a more descriptive naming for `USE_NEW_PROPLENGTH`

File: adapters/repos/db/priorityqueue/queue.go [hunk 7]
```diff
@@ -39,6 +39,19 @@ func NewMin[T supportedValueType](capacity int) *Queue[T] {
 	}
 }
 
+// NewMin constructs a priority queue which prioritizes items with smaller distance and smaller ID
+func NewMinWithId[T supportedValueType](capacity int) *Queue[T] {
+	return &Queue[T]{
+		items: make([]Item[T], 0, capacity),
+		less: func(items []Item[T], i, j int) bool {
+			if items[i].Dist == items[j].Dist {
+				return items[i].ID > items[j].ID
```
Comment (jeroiraz) [inline adapters/repos/db/priorityqueue/queue.go:48]: shouldn't it be `items[i].ID < items[j].ID`?

File: adapters/repos/db/shard_write_put.go (hunk 5 above)
Comment (amourao) [reply adapters/repos/db/shard_write_put.go:527]: This is related to the behavior when dealing with patching and partial updates, I'll add a comment to describe my reasoning.

File: adapters/repos/db/shard_write_inverted_lsm.go (hunk 6 above)
Comment (amourao) [reply adapters/repos/db/shard_write_inverted_lsm.go:93]: Yeah, I'm still debating on how to make this accessible.

File: adapters/repos/db/lsmkv/strategies_map.go (hunk 4 above)
Comment (amourao) [reply adapters/repos/db/lsmkv/strategies_map.go:391]: Agreed, I'll add more thorough  validation

File: adapters/repos/db/lsmkv/strategies_map.go (hunk 3 above)
Comment (amourao) [reply adapters/repos/db/lsmkv/strategies_map.go:291]: You are right, keys and values are always 8 bytes long, I'll add more validation

**Comment by amourao** [comment]
> ## [Quality Gate Failed](https://sonarcloud.io/dashboard?id=weaviate_weaviate&pullRequest=6176) **Quality Gate failed**
> 
> Failed conditions  [14.3% Duplication on New Code](https://sonarcloud.io/component_measures?id=weaviate_weaviate&pullRequest=6176&metric=new_duplicated_lines_density&view=list) (required ≤ 3%)
> 
> [See analysis details on SonarQube Cloud](https://sonarcloud.io/dashboard?id=weaviate_weaviate&pullRequest=6176)

The duplication here mostly comes from tests:
- `compactor_inverted_integration_test.go` for shares code with `compactor_map_integration_test.go`, but making it reusable would make it less readable;
- `bm25f_block_test.go` shares code with `bm25f_test.go`. Here it is even harder, as the index setup is the same, but the scores change slightly (by design) across tests.

**Review by parkerduckworth — APPROVED** [review]

**Pull request merged**